### DISCUSSION TITLE
Automatic speed detection ?

### DISCUSSION THREAD

**Discussion by HiFiPhile** [ataradov/usb-sniffer#21] [General]
Thanks for this awesome project !

As a [tinyusb](https://github.com/hathach/tinyusb) project maintainer, this analyzer is very interesting :)

I saw there is a speed detector implemented in FPGA but auto speed option is not exported in host side, is there any issue such as false detection ?

I built one and so far it works well for full-speed devices, high-speed devices stop enumerating once the cable longer than 1m.

I'm working on my hardware implantation trying to improve signal integrity https://github.com/HiFiPhile/usb-sniffer_hw/blob/master/USB_Sniffer.pdf :
- 4 Layer board (no need to save 2$ :) )
- 90 Ohm controlled impedance USB routing
- Minimize stub effect of comparators & phy

**Comment by ataradov** [comment]
PcapNG file format requires speed to be specified before capture starts. And overall automatic speed detection adds a lot of work and corner cases to the software without a real benefit.

It worked fine in limited testing I did.

**Comment by ataradov** [comment]
I also did some experiments with proposed changes. I added 510 Ohm resistors before the comparators. I cut the traces close to the connector pins on the existing board. I also drilled out the short layer jump on the DM line and replaced it with 0 Ohm resistor. I can't do anything with impedance matching on the existing board, of course.

None of that had any perceivable effect. 

But I also ran into some strange effect that I can't fully explain.  I use Lenovo ThinkCentre mini PC. It has USB ports on the front and in the back. I tried with USB stick directly plugged into the sniffer and an Android phone plugged using another USB cable.

At first I tried back USB for the data interface and front USB for the target interface. Back USB uses the same Chinese flexible cable. With flexible cable for the target interface, devices fail to enumerate. With a good cable, both devices work as expected and the capture works fine.

But then I tried to reverse the connectors using back USB as a target interface and the front USB as a data interface. With this change and both "bad" cables, it all worked as expected. 

As far as I can tell all of those USB connectors are attached to different ports of the same root hub. So, it looks like there are a lot of things stacking up to cause marginal behavior.  Routing of the USB connector inside the PC to the front panel, bad cables, losses introduced by the sniffer itself.

It is really hard to tell what exactly going on without proper equipment. But it does not look like there is a simple solution to this. Even improving impedance matching in the sniffer may not help, so using the best cables and trying different ports in case of issues may be necessary.

**Reply by a20dev** [reply]
> I also did some experiments with proposed changes. I added 510 Ohm resistors before the comparators. I cut the traces close to the connector pins on the existing board.

Have you tried with the traces to the comparators being cut (and without resistors)? The sniffer won't work, but does this make a difference for the DUT? That way we could rule out or confirm the stub to the comparators as culprit.

**Comment by HiFiPhile** [comment]
> None of that had any perceivable effect.

Oh that's sad :( 

I did a quick impedance calculation of your differential track which gives 112 Ohm, not too far compared to 90.

> So, it looks like there are a lot of things stacking up to cause marginal behavior. Routing of the USB connector inside the PC to the front panel, bad cables, losses introduced by the sniffer itself.

That's true. I had an old ASUS PC who use FPC cable for USB which is not great. In another Clevo one it has retimer ICs behind every ports.

Afterall without a high-end oscilloscope it's hard to know what happened.

**Comment by a20dev** [comment]
I computed about 150 Ω differential impedance (0.254 mm width, 0.254 mm distance, 0.35 µm copper, 1.55 mm dialectric with constant 4.6 for a 1.6 mm board).

**Reply by HiFiPhile** [reply]
Using different formula the results are different. I used Altium who has a 2D filed solver in most case it is not bad :

Diffrential-Coplaner: 0.19mm clearance 0.254 mm width, 0.254 mm distance, 0.35 µm copper, 1.5 mm dielectric Er 4.6

**Comment by a20dev** [comment]
The clearance is 0.19 mm only under the USB C connector. Elsewhere, the clearance is 0.254 mm.

**Reply by HiFiPhile** [reply]
No between USB track and GND is always the same, in fact it's 0.2mm not 0.19mm.
![image](https://github.com/ataradov/usb-sniffer/assets/4375114/5adad7eb-4c85-4dab-965d-e2b35cd4b34d)

**Comment by a20dev** [comment]
Yes, but the 90 Ω impedance to achieve is the differential impedance. We have a distance of 0.254 mm between the D+ and D- traces.  I didn't take the distance to the ground plane on the same layer into account as I haven't found a PCB impedance calculator that does that. Does your impedance calculator use all three distances (distance between layers, distance between the two traces, distances from a trace to the ground plane in the same layer)?

**Reply by HiFiPhile** [reply]
Yes, that's why I'm saying `Diffrential-Coplaner`

![image](https://github.com/ataradov/usb-sniffer/assets/4375114/f054c3b0-b394-42b7-adf7-78083c201175)

**Comment by a20dev** [comment]
The single-ended impedance should be above 45 Ω IIRC.

**Comment by a20dev** [comment]
I found a better online calculator which takes the distance to the reference plane in the same layer into account:
  https://www.protoexpress.com/tools/pcb-impedance-calculator/
About 118 Ω indeed.

**Comment by ataradov** [comment]
Tried a few more experiments:
1. Bare PCB with only connectors populated - both directly plugged in flash drive and a phone with extra cable work fine.
2. Remove 510 Ohm resistors from the existing real board - flash starts to work, phone still does not work.
3. Replace resistors with 1K - flash still works, phone still does not and line state detection still works too. 

So, it looks like both comparators and the PHY contribute to the behavior. And it may be possible to decouple the comparators with higher series resistance. Although it is hard to tell how marginal it would end up being anyway. I only tested with one USB flash drive.

PCB stubs on their own don't seem to matter and impedance mismatch is probably contributing something as well, but it it hard to tell how much.

And I don't think there is a way to better decouple the PHY without sacrificing the actual functionality. And I'm inclined to require better cabling setup rather than sacrifice the goal of the device. 

There is some info that TotalPhase sniffer works better on the same PC setup, but I don't think it is possible to compare them directly, since TotalPhase uses Type-B connectors, and there are not a lot of shoddy Type-B cables out there, they seem to have been designed before radical cost cutting and the cables still made are made with the same quality. And this sniffer works fine with good quality Type-C cables.

In all cases I only tried the "bad" cable, since I expect all combinations to work with the good cable.

**Comment by a20dev** [comment]
Maybe adding two op amps (e.g., one LTC6229 or two LTC6228 powered from 5 V rail) in unity-gain configuration might reduce the impact of the PHY and the comparators on the signal.

**Comment by ataradov** [comment]
If someone wants to experiment with that - go for it. I don't really care enough to bother doing that. I don't see a real issue here, just use good cables and it all works. Adding extra components will just complicate the design for no benefit.

**Reply by HiFiPhile** [reply]
Am I understood correctly that high-speed terminal resistor can't be disabled ?

![image](https://github.com/ataradov/usb-sniffer/assets/4375114/9b384d48-bbf7-47ce-a321-762432ccb9e1)

**Comment by ataradov** [comment]
In a non-driving mode all termination should be disabled. Otherwise nothing would work at all.

The whole point of the comparators is to see when one or both devices enable HS termination. And you can clearly see that in the log. You would not see that if this PHY already enabled termination on its own.
